Framework/DataHandling/src/LoadEMU.cpp

```diff
@@ -1238,22 +1238,41 @@ void LoadEMUHdf::exec() {
   // dataset index to be loaded
   m_datasetIndex = Base::getProperty(SelectDatasetStr);
 
-  // if path provided build the file path
+  // if path provided build the file path from the directory name and dataset
+  // number from the hdf file, however if this is not a valid path then try
+  // the basename with a '.bin' extension
   if (fs::is_directory(evtPath)) {
     NeXus::NXRoot root(hdfFile);
     NeXus::NXEntry entry = root.openFirstEntry();
     auto eventDir = GetNeXusValue<std::string>(entry, "instrument/detector/daq_dirname", "./", 0);
     auto dataset = GetNeXusValue<int32_t>(entry, "instrument/detector/dataset_number", 0, m_datasetIndex);
+    if (dataset < 0) {
+      std::string message("Negative dataset index recorded in HDF, reset to zero!");
+      g_log.error(message);
+      dataset = 0;
+    }
 
-    // build the path to the event file as if a relative or absolute
-    // path is passed:
-    //   'relpath/[daq_dirname]/DATASET_[n]/EOS.bin' or the
-    char buffer[255] = {};
-    snprintf(buffer, sizeof(buffer), "%s/DATASET_%d/EOS.bin", eventDir.c_str(), dataset);
-    fs::path path = evtPath;
-    path /= buffer;
-    path = fs::absolute(path);
-    evtPath = path.generic_string();
+    // build the path to the event file using the standard storage convention at ansto:
+    //   'relpath/[daq_dirname]/DATASET_[n]/EOS.bin'
+    // but if the file is missing, try relpath/{source}.bin
+    std::stringstream buffer;
+    buffer << eventDir.c_str() << "/DATASET_" << dataset << "/EOS.bin";
+    fs::path filePath = evtPath;
+    filePath /= buffer.str();
+    filePath = fs::absolute(filePath);
+    std::string nomPath = filePath.generic_string();
+    if (fs::is_regular_file(nomPath)) {
+      evtPath = nomPath;
+    } else {
+      fs::path hp = hdfFile;
+      buffer.str("");
+      buffer.clear();
+      buffer << hp.stem().generic_string().c_str() << ".bin";
+      fs::path path = evtPath;
+      path /= buffer.str();
+      path = fs::absolute(path);
+      evtPath = path.generic_string();
+    }
```

Comment:
You can  use the `/` operator in `boost::filesystem::path` to concatenate different sections and also `std::string` allows the `+` operator to concatenate strings, and you can use `std::to_string` to convert a lot of datatypes to string so you don't need to use the `stringstream`. Also, `is_regular_file` takes a path object so you don't need the generic string, and you can use `replace_extension` instead of handling the string.

So I think this block could be changed to something like:

```c++
fs::path filePath = fs::absolute(fs::path(evtPath) / eventDir /
                    ("DATASET_" + std::to_string(dataset)) / "EOS.bin");
if (!fs::is_regular_file(filePath)) {
  path = fs::absolute(evtPath / fs::path(hdfFile).replace_extension(".bin"));
}
evtPath = filePath.generic_string();
```

Reply:
I've adopted your suggestion, thanks.